typing/typecore.ml

```diff
@@ -5132,14 +5165,56 @@ and type_application env app_loc expected_mode funct funct_mode sargs =
       in
       submode ~loc:app_loc ~env
         (Value_mode.of_alloc mres) expected_mode;
-      let marg = argument_mode marg in
+      let marg =
+        mode_argument ~position:expected_mode.position ~partial_app:false marg
+      in
       let exp = type_expect env marg sarg (mk_expected ty_arg) in
       check_partial_application false exp;
-      ([Nolabel, Some exp], ty_res)
+      ([Nolabel, Arg exp], ty_res)
   | _ ->
-    let ty = funct.exp_type in
-    type_args [] ty (instance ty) sargs
-      (Value_mode.regional_to_global_alloc funct_mode)
+      let ty = funct.exp_type in
+      let ignore_labels =
+        !Clflags.classic ||
+        begin
+          let ls, tvar = list_labels env funct.exp_type in
+          not tvar &&
+          let labels = List.filter (fun l -> not (is_optional l)) ls in
+          List.length labels = List.length sargs &&
+          List.for_all (fun (l,_) -> l = Nolabel) sargs &&
+          List.exists (fun l -> l <> Nolabel) labels &&
+          (Location.prerr_warning
+             funct.exp_loc
+             (Warnings.Labels_omitted
+                (List.map Printtyp.string_of_label
+                          (List.filter ((<>) Nolabel) labels)));
+           true)
+        end
+      in
+      let ty_ret, mode_ret, rev_args =
+        collect_apply_args env funct ignore_labels ty (instance ty)
+          (Value_mode.regional_to_global_alloc funct_mode) sargs
+      in
+      let ty_ret, mode_ret, closed_args, rev_args =
+        type_omitted_parameters ty_ret mode_ret rev_args
+      in
+      let position = expected_mode.position in
+      let partial_app = is_partial_apply rev_args in
+      let args =
+        (* Careful: the order matters here. Using [List.rev_map] would be
+           incorrect. *)
+        List.map (fun arg ->
+          type_apply_arg env ~position ~partial_app arg)
+          (List.rev rev_args)
+      in
+      List.iter
+        (fun (marg, sarg) ->
+           submode ~loc:sarg.pexp_loc ~env
+             (Value_mode.of_alloc marg)
+             (mode_partial_application expected_mode))
+        closed_args;
```

Comment:
I think it would be cleaner to do this in `type_omitted_parameters`, at the point where `closed_args` grows in the `Omitted` case. That way, you can remove `closed_args` entirely.